src/Microsoft.OpenApi/Validations/Rules/OpenApiComponentsRules.cs

```diff
@@ -17,11 +17,11 @@ public static class OpenApiComponentsRules
         /// <summary>
         /// The key regex.
         /// </summary>
-        public static Regex KeyRegex = new Regex(@"^[a-zA-Z0-9\.\-_]+$");
+        public static Regex KeyRegex = new Regex(@"^[a-zA-Z0-9\.\-_\[\]]+$");
 
         /// <summary>
         /// All the fixed fields declared above are objects
-        /// that MUST use keys that match the regular expression: ^[a-zA-Z0-9\.\-_]+$.
+        /// that MUST use keys that match the regular expression: ^[a-zA-Z0-9\.\-_\[\]]+$.
         /// </summary>
         public static ValidationRule<OpenApiComponents> KeyMustBeRegularExpression =>
             new ValidationRule<OpenApiComponents>(
```

Comment:
The existing RegEx is a valid rule in V3.

For now, we should fix it in a different way. Create a new rule in your project - and when you call the reader, specify a ruleset with this new rule instead of the default V3 rule.

There was a discussion earlier about creating rulesets for V2 and V3, but I think we decided against it at that point. It seems like this is creating more work for our customers though @darrelmiller ?

Reply:
I think we should translate invalid characters into some valid character.  e.g. into a '-'.  Having square brackets in a component name would break JSON pointer references as square brackets have a specific meaning in a JSON pointer.

Reply:
@darrelmiller I thought the only special characters in JSON pointer are '/' and '~'. Regardless, I think the rule here should stay the same since that's what's required in V3. I would be a bit cautious replacing characters directly since that might cause unexpected issues for our customers - throwing validation error seems more predictable to me.

Regardless, we need to allow people who's using the DOM for their V2 document to not get blocked by this rule. Currently they can manually remove the rule, but that seems like an unnecessary and a bit obscure extra work. What do you think about creating a pre-defined rulesets for V2 / V3 / Vx that the customers can choose from?

---
In reply to: [187430025](https://github.com/Microsoft/OpenAPI.NET/pull/254#discussion_r187430025) [](ancestors = 187430025)

Reply:
@PerthCharern You are right about the special characters.  I always get mixed up about how you reference array elements with JSON Pointer.  
I understand the concern about changing characters, but I'm concerned about allowing V2 documents to be imported and then exported as V3 but with invalid content.  The DOM is a V3 object model.  It is going to make changes to V2 imported documents.  It changes how produces/consumes works.  Round tripping V2 to V2 wasn't a primary goal.  Upgrading V2 to V3 and downgrading V3 to V2 was.  Roundtripping V2 was a bit of an accidental benefit.

Customers understand that when translating from one version to the other there is going to be possible changes to the doc.  If we create V2 ruleset, we are implying that we "support V2 documents" and I think that is a bit of a stretch.

Reply:
Darrel, then should openapi.net replace all unsupported chars? or its going to be selective of replacing the one's that are supported in V2 but not in V3?

---
In reply to: [187445907](https://github.com/Microsoft/OpenAPI.NET/pull/254#discussion_r187445907) [](ancestors = 187445907)

Reply:
@Shwetap05 That's an interesting question.  I think we should only replace chars during the deserialization of V2 documents.  V2 had no limit on the allowed characters, so all unsupported characters should be changed.
However, if someone is importing a V3 document, we should not replace unsupported characters, we should create an error.
This is because we want to be explicit that we are "converting" a V2 to a V3 when we import so some transformation is expected.  If someone is importing a V3, we shouldn't change anything.

Reply:
@darrelmiller @Shwetap05  I agree with Darrel's approach above.

Reply:
To be clear, I think this is how that approach translates to the implementation:

- keep this rule the same (This will ensure that the final DOM object still has the correct set of characters.)

- convert unsupported characters in V2 deserializer.